### PR TITLE
fix #22548;environment misses for type reference in iterator access n…

### LINKED ISSUE #21737: internal error: environment misses when using toSeq on iterator, inside nested iterator
### Description

Got the error when using glob library, here is the minimum reproducible code. `mytoSeq` is striped version of standard toSeq template.
```nim
template mytoSeq*(iter: untyped): untyped =
  var result: seq[typeof(iter)]# = @[]
  for x in iter:
    result.add(x)
  result

iterator test(dir:int):int=
  yield 1234

iterator walkGlobKinds (): int =
  let dir = 123
  let it = mytoSeq(test(dir))  

proc main()=
    let it = iterator():int=
      for path in walkGlobKinds():
          yield path
main()
```

### Nim Version

test on both 
```
Nim Compiler Version 1.6.12 [Linux: amd64]
Compiled at 2023-03-10
Copyright (c) 2006-2023 by Andreas Rumpf

git hash: 1aa9273640c0c51486cf3a7b67282fe58f360e91
active boot switches: -d:release
```
and
```
Nim Compiler Version 1.9.3 [Linux: amd64]
Compiled at 2023-04-27
Copyright (c) 2006-2023 by Andreas Rumpf

git hash: 220b45048983998675df761d4f33cd31128f10d5
active boot switches: -d:release
```

### Current Output

```text
Error: internal error: environment misses: dir
```

### Expected Output

```text
Compile fine
```

### Possible Solution

_No response_

### Additional Information

_No response_

### LINKED ISSUE #22548: "environment misses" for type reference in iterator access nested in closure
### Description

```nim
type Xxx[T] = object

iterator x(v: string): char =
  var v: Xxx[int]

  var y: v.T

  echo y

proc bbb(vv: string): proc () =
  proc xxx() =
    for c in x(vv):
      echo c

  return xxx

bbb("test")()
```

### Nim Version

1.6.14

### Current Output

```
testit.nim(9, 10) Error: internal error: environment misses: v
```

### Expected Output

_No response_

### Possible Solution

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by bung87**
…ested in closure

fix #22548
fix https://github.com/nim-lang/Nim/issues/21737

### CODE REVIEW

**Comment by bung87** [comment]
!nim c
```nim
template mytoSeq*(iter: untyped): untyped =
  var result: seq[typeof(iter)]# = @[]
  for x in iter:
    result.add(x)
  result

iterator test(dir:int):int=
  yield 1234

iterator walkGlobKinds (): int =
  let dir = 123
  let it = mytoSeq(test(dir))  

proc main()=
    let it = iterator():int=
      for path in walkGlobKinds():
          yield path
main()
```

refs: #21737

**Comment by juancarlospaco** [comment]
@bung87 
Comment must start with the `!nim ` because it checks the first line only for performance for fast builds, 
you can have any arbitrary text and links after the code blocks (or write the text on other comment): 
https://github.com/juancarlospaco/nimrun-action/blob/ab65b08c2b52e338f1069de69f51cc4f5253dfb5/index.js#L187

**Comment by bung87** [comment]
!nim c
```nim
template mytoSeq*(iter: untyped): untyped =
  var result: seq[typeof(iter)]# = @[]
  for x in iter:
    result.add(x)
  result

iterator test(dir:int):int=
  yield 1234

iterator walkGlobKinds (): int =
  let dir = 123
  let it = mytoSeq(test(dir))  

proc main()=
    let it = iterator():int=
      for path in walkGlobKinds():
          yield path
main()
```

refs: #21737

File: compiler/lambdalifting.nim [hunk 1]
```diff
@@ -516,6 +525,12 @@ proc detectCapturedVars(n: PNode; owner: PSym; c: var DetectionPass) =
       detectCapturedVars(n[namePos], owner, c)
   of nkReturnStmt:
     detectCapturedVars(n[0], owner, c)
+  of nkIdentDefs:
+    if not containsCallKinds(n[1]):
+      detectCapturedVars(n[^1], owner, c)
```
Comment (Araq) [inline compiler/lambdalifting.nim]: Why not always simply do `detectCapturedVars(n[^1], owner, c)`?

Comment (bung87) [reply compiler/lambdalifting.nim]: it will fails on `zero-functional`'s   `gensym4090: seq[type(_ZF_iterType()[1])]`

Comment (Araq) [reply compiler/lambdalifting.nim]: How so? `n[^1]` is the `value` part of `var v: T = value` and it is empty for your snippet.

Comment (bung87) [reply compiler/lambdalifting.nim]: so that it avoid `detectCapturedVars` of `_ZF_iterType` then it loss env

Comment (Araq) [reply compiler/lambdalifting.nim]: You misunderstand me. I'm asking why the patch is not just:

```suggestion
  of nkIdentDefs:
    detectCapturedVars(n[^1], owner, c)
```

Without the `if`.

Comment (bung87) [reply compiler/lambdalifting.nim]: I tried at first time, it looks much cleaner and will solve https://github.com/nim-lang/Nim/issues/21737, but `detectCapturedVars` also calls `addClosureParam`, if we avoid processing `detectCapturedVars` on `gensym4090: seq[type(_ZF_iterType()[1])]` (the type modifier), we will loss env param for `_ZF_iterType`

Comment (Araq) [reply compiler/lambdalifting.nim]: Makes no sense to me. So we need some special case code for the `typ` position of the variable declaration then but that's it.

Comment (bung87) [reply compiler/lambdalifting.nim]: I tried to add extra code handle all cases, but no luck, https://github.com/nim-lang/Nim/issues/21737 result to `cgen` error, others passed.

File: compiler/lambdalifting.nim [hunk 2]
```diff
@@ -417,6 +418,9 @@ Consider:
 
 """
 
+proc isTypeOf(ident: PIdent; cache: IdentCache): bool =
+  ident == getIdent(cache, "typeof") or ident == getIdent(cache, "type")
+
```
Comment (Araq) [inline compiler/lambdalifting.nim:423]: Why does the frontend not transform these into `nkType` nodes? It should and would make this far easier.

Comment (bung87) [reply compiler/lambdalifting.nim:423]: I was expecting these are just `nkTypeOfExpr`

Comment (Araq) [reply compiler/lambdalifting.nim:423]: Yes, even better. :-)

Comment (bung87) [reply compiler/lambdalifting.nim:423]: what's the node processing path goes to here?

Comment (Araq) [reply compiler/lambdalifting.nim:423]: `semMagicAfterOverloadResolution` should turn the magic mType and mTypeOf to nkTypeOfExpr

Comment (bung87) [reply compiler/lambdalifting.nim:423]: tried , didn't found chance transform the node , now only detect via magic, seems sufficient

**Comment by yglukhov** [comment]
@bung87 this pr has broken my hacky unofficial [yasync](https://github.com/yglukhov/yasync) experiments, because after this pr the fields in the env objects are in different order from their declaration in the iter, and that's the behavior I relied on. Is this new behavior intentional? Can it be circumvented somehow?

**Comment by bung87** [comment]
make `detectCapturedVars` also processing on the symbol

```nim
of nkIdentDefs:
    detectCapturedVars(n[0], owner, c)
    detectCapturedVars(n[^1], owner, c)
```

**Comment by yglukhov** [comment]
@bung87 sorry I didn't understand. Does it mean you'll be looking into this? :)

**Comment by bung87** [comment]
I've tested locally, `compiler/lambdalifting.nim` line start from 526, change to the solution above. but I will not make PR

**Comment by yglukhov** [comment]
Indeed, it seems to work, thanks. I'll make a PR now.

**Pull request merged**